apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.SessionPoolKey.cs

```diff
@@ -0,0 +1,88 @@
+        /// </summary>
+        public DatabaseName DatabaseName { get; }
+
+        /// <summary>
+        /// The database role of the sessions pooled on the associated session pool.
+        /// </summary>
+        public string DatabaseRole { get; }
+
+        /// <summary>
+        ///  Instantiates a new instance of the <see cref="SessionPoolKey"/> object.
+        /// </summary>
+        /// <param name="databaseName">The database name for the associated pool. Must not be null</param>
+        /// <param name="databaseRole">The database role for the associated pool. May be null
+        /// in which case the default (public) database role is used if fine grained access control
+        /// is enabled; otherwise the database role is ignored.</param>
+        private SessionPoolKey(DatabaseName databaseName, string databaseRole)
+        {
+            DatabaseName = GaxPreconditions.CheckNotNull(databaseName, nameof(databaseName));
+            DatabaseRole = databaseRole;
+        }
+
+        /// <summary>
+        /// Creates a <see cref="SessionPoolKey"/> with the given database name. 
+        /// </summary>
+        /// <param name="databaseName">The database name for the associated pool. Must not be null</param>
+        /// <returns></returns>
+        public static SessionPoolKey Create(DatabaseName databaseName) =>
+            new SessionPoolKey(databaseName, null);
+
+        /// <summary>
+        /// Creates a <see cref="SessionPoolKey"/> with the given database role. 
+        /// </summary>
+        /// <param name="databaseRole">The database role for the associated pool. May be null
+        /// in which case the default (public) database role is used if fine grained access control
+        /// is enabled; otherwise the database role is ignored.</param>
+        /// <returns></returns>
+        public SessionPoolKey WithDatabaseRole(string databaseRole) =>
+            new SessionPoolKey(DatabaseName, databaseRole);
+
+        /// <inheritdoc/>
+        public override string ToString() =>
+            $"DatabaseName: {DatabaseName}; DatabaseRole: {DatabaseRole}";
+
+        /// <inheritdoc/>
+        public override int GetHashCode() =>
+            DatabaseRole is null ? DatabaseName.GetHashCode() :
+            GaxEqualityHelpers.CombineHashCodes(DatabaseRole.GetHashCode(), DatabaseName.GetHashCode());
+
+        /// <inheritdoc/>
+        public override bool Equals(object obj) =>
+            Equals(obj as SessionPoolKey);
+
+        /// <summary> 
+        /// Determines whether <paramref name="other"/> is equal to this instance.
+        /// </summary>
+        /// <returns><c>true</c> if the specified object is equal to this instance; otherwise <c>false</c>.</returns>
+        public bool Equals(SessionPoolKey other) =>
+            (other is null) ? false : DatabaseName.Equals(other.DatabaseName) && DatabaseRole == other.DatabaseRole;
```

Comment:
Change the `.Equals` to `==` for simplicity.

Reply:
Oh, I asked for this to be Equals, do resource names redefine the equality operators? Note this is not the string, but the resource name?

Reply:
I hadn't noticed that it's a DatabaseName. But yes, we *do* overload equality operators. See https://cloud.google.com/dotnet/docs/reference/Google.Cloud.Spanner.Common.V1/latest/Google.Cloud.Spanner.Common.V1.DatabaseName#Google_Cloud_Spanner_Common_V1_DatabaseName_op_Equality_Google_Cloud_Spanner_Common_V1_DatabaseName_Google_Cloud_Spanner_Common_V1_DatabaseName_